File: src/chapter1_7.rs
//! Write an algorithm such that if an element in an MxN matrix is 0, its entire row and column is set to 0.

/// Flag rows and columns then clear data.
#[allow(dead_code)]
fn set_zeros(matrix: &mut Vec<Vec<i32>>) {
    // does not check that the dimensions are correct
    let rows = matrix.len();
    let cols = matrix[0].len();
    let mut row_flag = vec![false; rows];
    let mut col_flag = vec![false; cols];

    // flag rows and columns
    for i in 0..rows {
        for j in 0..cols {
            if matrix[i][j] == 0 {
                row_flag[i] = true;
                col_flag[j] = true;
            }
        }
    }

    // clear data
    for i in 0..rows {
        for j in 0..cols {
            if row_flag[i] || col_flag[j] {
                matrix[i][j] = 0;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_algo() {
        let mut matrix = vec![
            vec![1, 1, 1, 1],
            vec![1, 0, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
        ];
        let expected = vec![
            vec![1, 0, 1, 1],
            vec![0, 0, 0, 0],
            vec![1, 0, 1, 1],
            vec![1, 0, 1, 1],
            vec![1, 0, 1, 1],
            vec![1, 0, 1, 1],
        ];
        set_zeros(&mut matrix);
        assert_eq!(matrix, expected);


        let mut matrix = vec![
            vec![1, 1, 1, 1],
            vec![1, 0, 0, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 0, 1],
            vec![0, 1, 1, 1],
            vec![1, 1, 1, 1],
        ];
        let expected = vec![
            vec![0, 0, 0, 1],
            vec![0, 0, 0, 0],
            vec![0, 0, 0, 1],
            vec![0, 0, 0, 0],
            vec![0, 0, 0, 0],
            vec![0, 0, 0, 1],
        ];
        set_zeros(&mut matrix);
        assert_eq!(matrix, expected);

        let mut matrix = vec![
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
        ];
        let expected = vec![
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
            vec![1, 1, 1, 1],
        ];
        set_zeros(&mut matrix);
        assert_eq!(matrix, expected);
    }
}

File: elixir/lib/chapter1_7.ex
defmodule Crack.Chapter1_7 do
  @moduledoc """
  1.7 Matrix 0ing:

  Write an algorithm such that if an element of an MxN matrix is 0, its entire
  row and column is set to 0.
  """

  @doc """
  Turn rows and columns contains 0s into 0s.

      iex> matrix = [
      ...>   [1, 1, 1, 0],
      ...>   [1, 1, 1, 1],
      ...>   [1, 1, 1, 1],
      ...>   [0, 1, 1, 1]]
      [[1, 1, 1, 0], [1, 1, 1, 1], [1, 1, 1, 1], [0, 1, 1, 1]]
      iex> Crack.Chapter1_7.start(matrix)
      [[0, 0, 0, 0],
       [0, 1, 1, 0],
       [0, 1, 1, 0],
       [0, 0, 0, 0]]

       iex> matrix = [[0]]
       [[0]]
       iex> Crack.Chapter1_7.start(matrix)
       [[0]]

       iex> matrix = [[1]]
       [[1]]
       iex> Crack.Chapter1_7.start(matrix)
       [[1]]

       iex> matrix = [
       ...>   [1, 1, 1, 1, 1],
       ...>   [1, 1, 1, 1, 1],
       ...>   [1, 1, 0, 1, 1],
       ...>   [1, 1, 1, 1, 1],
       ...>   [1, 1, 1, 1, 1]]
       [[1, 1, 1, 1, 1],[1, 1, 1, 1, 1], [1, 1, 0, 1, 1], [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1]]
       iex> Crack.Chapter1_7.start(matrix)
       [[1, 1, 0, 1, 1],
        [1, 1, 0, 1, 1],
        [0, 0, 0, 0, 0],
        [1, 1, 0, 1, 1],
        [1, 1, 0, 1, 1]]

        iex> matrix = []
        []
        iex> Crack.Chapter1_7.start(matrix)
        []

        iex> matrix = [[1, 1, 1, 0, 1]]
        [[1, 1, 1, 0, 1]]
        iex> Crack.Chapter1_7.start(matrix)
        [[0, 0, 0, 0, 0]]

        iex> matrix = [
        ...>   [1, 1],
        ...>   [1, 1],
        ...>   [1, 0],
        ...>   [1, 1],
        ...>   [0, 1]]
        [[1, 1], [1, 1], [1, 0], [1, 1], [0, 1]]
        iex> Crack.Chapter1_7.start(matrix)
        [[0, 0],
         [0, 0],
         [0, 0],
         [0, 0],
         [0, 0]]


  > Note this is probably absolutely terrible in terms of performance. But I
    quite like this solution for some reason.

  """
  def start(matrix) do
    zero_rows =
      matrix
      |> Enum.map(&contains_zeros?/1)

    zero_cols =
      matrix
      |> transpose()
      |> Enum.map(&contains_zeros?/1)

    matrix
    |> Enum.zip(zero_rows)
    |> Enum.map(&replace_zeros/1)
    |> transpose()
    |> Enum.zip(zero_cols)
    |> Enum.map(&replace_zeros/1)
    |> transpose()
  end

  defp transpose(matrix) do
    matrix
    |> Enum.zip()
    |> Enum.map(&Tuple.to_list/1)
  end

  defp contains_zeros?(list) do
    Enum.any?(list, &(&1 === 0))
  end

  defp replace_zeros({list, false}), do: list

  defp replace_zeros({[], true}), do: []

  defp replace_zeros({[_ | tail], true}), do: [0 | replace_zeros({tail, true})]
end
